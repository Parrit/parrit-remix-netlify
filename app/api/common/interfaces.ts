export interface LoginRequest {
  projectName?: string;
  password?: string;
}

export interface LoginResult {
  message: string;
}

export interface Project {
  id: string;
  name: string;
  pairingBoards: PairingBoard[];
  people: Person[];
}

export interface PairingBoard {
  id: string;
  name: string;
  exempt: boolean;
  people: Person[];
  roles: Role[];
}

export interface Person {
  id: string;
  name: string;
}

export interface Role {
  id: string;
  name: string;
}

export interface PairingArrangementDTO {
  id: number; // this should be changed to string
  pairingTime: string;
  pairingHistories: PairingHistory[];
}

export interface PairingHistory {
  pairingBoardName: string;
  people: Person[];
}
