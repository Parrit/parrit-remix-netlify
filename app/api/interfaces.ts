export interface FormErrors<T> {
  server?: string;
  fields?: Record<keyof T, string | null>;
  status: 400 | 401 | 500;
}

export const isFormErrors = <T>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  arg: FormErrors<T> | any
): arg is FormErrors<T> => {
  return (arg as FormErrors<T>).status > 0;
};
