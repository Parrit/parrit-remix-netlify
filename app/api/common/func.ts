import {
  FLOATING_IDX,
  PairingBoard,
  Person,
  Project,
} from "./interfaces/parrit.interfaces";
import * as R from "ramda";

// Functional programing utilities for working with interfaces

export const movePerson = (
  project: Project,
  person: Person,
  position: PairingBoard
) =>
  R.pipe<[Project], Project, Project>(
    removePerson(person),
    addPerson(person, position)
  )(project);

export const addPerson = R.curry(
  (person: Person, position: PairingBoard, project: Project) => {
    const f1 = addPersonToPosition(person, position);
    const f2 =
      position.id === FLOATING_IDX ? addPersonToFloating(person) : R.identity;
    return R.pipe<[Project], Project, Project>(f1, f2)(project);
  }
);

const addPersonToPosition = R.curry(
  (person: Person, position: PairingBoard, project: Project) =>
    R.over<Project, PairingBoard[]>(
      R.lensProp("pairingBoards"),
      R.map<PairingBoard, PairingBoard>((x) => {
        if (x.id === position.id) {
          return R.over<PairingBoard, Person[]>(
            R.lensProp("people"),
            R.append({ ...person, pairing_board_id: position.id }),
            x
          );
        }
        return x;
      }),
      project
    )
);

const addPersonToFloating = R.curry((person: Person, project: Project) =>
  R.over<Project, Person[]>(
    R.lensPath(["floating", "people"]),
    R.append(person),
    project
  )
);

export const removePerson = R.curry(
  (person: Person, project: Project): Project =>
    R.pipe<[Project], Project, Project>(
      removePersonFromBoards(person),
      removePersonFromFloating(person)
    )(project)
);

const removePersonFromBoards = R.curry(
  (person: Person, project: Project): Project =>
    R.over<Project, PairingBoard[]>(
      R.lensProp("pairingBoards"),
      R.map<PairingBoard, PairingBoard>((x) => {
        if (x.people.includes(person)) {
          return R.set<PairingBoard, Person[]>(
            R.lensProp("people"),
            R.without(x.people, [person]),
            x
          );
        }
        return x;
      }),
      project
    )
);

const removePersonFromFloating = R.curry(
  (person: Person, project: Project): Project =>
    R.set<Project, Person[]>(
      R.lensPath(["floating", "people"]),
      R.without([person], project.floating.people),
      project
    )
);

const pbIndexById = R.curry((pbs: PairingBoard[], pb_id: string) =>
  R.findIndex<PairingBoard>((pb) => pb.id === pb_id)(pbs)
);

const renamePairingBoardFromList = R.curry(
  (pb_id: string, rename: string, pbs: PairingBoard[]) => {
    const foundIndex = pbIndexById(pbs, pb_id);
    if (foundIndex < 0) {
      return pbs;
    }
    return R.adjust<PairingBoard>(foundIndex, R.assoc("name", rename), pbs);
  }
);

export const renamePairingBoard = R.curry(
  (project: Project, pb_id: string, rename: string): Project =>
    R.over<Project, PairingBoard[]>(
      R.lensProp("pairingBoards"),
      renamePairingBoardFromList(pb_id, rename),
      project
    )
);

export const canAPairingBeMade = (project: Project): boolean => {
  const atLeast2Floaters = project.floating.people.length >= 2;
  const atLeast1Floater = project.floating.people.length >= 1;
  const atLeast1EmptyBoard = getEmptyPairingBoard(project) !== undefined;
  const atLeast1UnpairedSticker = unpairedStickingPeople(project).length >= 1;
  return (
    (atLeast1Floater && atLeast1UnpairedSticker) ||
    (atLeast2Floaters && atLeast1EmptyBoard)
  );
};

export const getEmptyPairingBoard = (
  project: Project
): PairingBoard | undefined => {
  return project.pairingBoards.find(
    (pb) => pb.people.length === 0 && !pb.exempt
  );
};

const unpairedStickingPeople = (project: Project): Person[] =>
  project.pairingBoards
    .filter((board) => !board.exempt)
    .filter((board) => board.people.length === 1)
    .flatMap((board) => board.people);

const findPairingBoardByPerson = (
  project: Project,
  person: Person
): PairingBoard | undefined =>
  project.pairingBoards.find((pb) => pb.people.find((p) => p.id === person.id));
