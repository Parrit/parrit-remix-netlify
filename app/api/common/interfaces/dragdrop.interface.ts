export type DragType = "Person" | "Role";

export enum DropType {
  Floating = "Floating",
  PairingBoard = "PairingBoard",
  TrashBin = "TrashBin",
}

export interface DragItem {
  id: string;
  name: string;
  type: DragType;
  pairing_board_id: string;
}

export interface DropItem {
  type: DropType;
}

export interface DropTarget {
  isOver: boolean;
  canDrop: boolean;
}

export interface DragResult {}
