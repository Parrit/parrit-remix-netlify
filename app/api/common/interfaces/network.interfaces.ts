export interface LoginRequest {
  projectName?: string;
  password?: string;
}

export interface LoginResult {
  message: string;
}
