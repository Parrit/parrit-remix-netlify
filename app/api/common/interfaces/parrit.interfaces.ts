import { DragItem } from "~/api/common/interfaces/dragdrop.interface";

export interface Project {
  id: string;
  name: string;
  pairingBoards: PairingBoard[];
  floating: PairingBoard;
  people: Person[]; // the new cannonical location for people
  roles: Role[]; // the new cannonical location for roles
}

export const FLOATING_IDX = "-1";

export interface PairingBoard {
  id: string;
  name: string;
  exempt: boolean;
}

export interface Person extends DragItem {
  id: string;
  name: string;
  type: "Person";
}

export interface Role extends DragItem {
  id: string;
  name: string;
  type: "Role";
}

export interface ProjectPairingSnapshot {
  id: string;
  projectId: string;
  pairingTime: string;
  pairingInstances: PairingInstance[];
}

export interface PairingInstance {
  projectId: string;
  pairingBoardName: string | null;
  people: Person[];
  pairingTime: string;
}
