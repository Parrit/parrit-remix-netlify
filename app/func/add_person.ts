import {
  curry,
  identity,
  pipe,
  over,
  lensProp,
  map,
  append,
  lensPath,
} from "ramda";
import {
  Person,
  PairingBoard,
  Project,
  FLOATING_IDX,
} from "~/api/common/interfaces/parrit.interfaces";

export default curry(
  (person: Person, position: PairingBoard, project: Project) => {
    const f1 = add_person_to_position(person, position);
    const f2 =
      position.id === FLOATING_IDX ? add_person_to_floating(person) : identity;
    return pipe<[Project], Project, Project>(f1, f2)(project);
  }
);

const add_person_to_position = curry(
  (person: Person, position: PairingBoard, project: Project) =>
    over<Project, PairingBoard[]>(
      lensProp("pairingBoards"),
      map<PairingBoard, PairingBoard>((x) => {
        if (x.id === position.id) {
          return over<PairingBoard, Person[]>(
            lensProp("people"),
            append({ ...person, pairing_board_id: position.id }),
            x
          );
        }
        return x;
      }),
      project
    )
);

const add_person_to_floating = curry((person: Person, project: Project) =>
  over<Project, Person[]>(
    lensPath(["floating", "people"]),
    append(person),
    project
  )
);
