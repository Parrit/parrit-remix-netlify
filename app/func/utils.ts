/**
 * These files are tested via the functionality of other tests.
 * That's what makes them utils and not their own functions
 */

import {
  Project,
  Person,
  PairingBoard,
} from "~/api/common/interfaces/parrit.interfaces";

export const can_a_pairing_be_made = (project: Project): boolean => {
  const atLeast2Floaters = project.floating.people.length >= 2;
  const atLeast1Floater = project.floating.people.length >= 1;
  const atLeast1EmptyBoard = get_empty_pairing_board(project) !== undefined;
  const atLeast1UnpairedSticker = unpaired_sticking_people(project).length >= 1;
  return (
    (atLeast1Floater && atLeast1UnpairedSticker) ||
    (atLeast2Floaters && atLeast1EmptyBoard)
  );
};

export const unpaired_sticking_people = (project: Project): Person[] =>
  project.pairingBoards
    .filter((board) => !board.exempt)
    .filter((board) => board.people.length === 1)
    .flatMap((board) => board.people);

export const get_empty_pairing_board = (
  project: Project
): PairingBoard | undefined => {
  return project.pairingBoards.find(
    (pb) => pb.people.length === 0 && !pb.exempt
  );
};

export const all_people_in_project = (project: Project): Person[] => {
  const people: Person[] = [...project.floating.people];
  project.pairingBoards.forEach((pb) =>
    pb.people.forEach((person) => people.push(person))
  );
  return people;
};

export const find_pairing_board_by_person = (
  project: Project,
  person: Person
): PairingBoard | undefined =>
  project.pairingBoards.find((pb) => pb.people.find((p) => p.id === person.id));
