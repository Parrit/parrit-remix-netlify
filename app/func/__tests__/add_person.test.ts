import { describe, expect, it } from "@jest/globals";
import add_person from "../add_person";
import { FRIZZLE, PB_RED, PROJECT_START } from "./common";
import { can_a_pairing_be_made, floating_people } from "../utils";
import { FLOATING_IDX } from "~/api/common/interfaces/parrit.interfaces";

describe("add_person", () => {
  it("adding to a pb", () => {
    const result = add_person(FRIZZLE, PB_RED.id, PROJECT_START);
    expect(
      result.people.filter(
        ({ pairing_board_id }) => pairing_board_id === PB_RED.id
      )
    ).toStrictEqual([
      {
        id: "red_parrit_id",
        name: "red_parrit",
        pairing_board_id: "red_id",
        project_id: "20232",
        type: "Person",
      },
      {
        id: "99",
        name: "Ms. Frizzle",
        pairing_board_id: "red_id",
        project_id: "20232",
        type: "Person",
      },
    ]);
    expect(floating_people(result).length).toBe(
      floating_people(PROJECT_START).length
    );
    expect(can_a_pairing_be_made(result)).toBe(false); // only one floater, all others are paired
  });

  it("add to floating", () => {
    const start_count = floating_people(PROJECT_START).length;
    const result = add_person(FRIZZLE, FLOATING_IDX, PROJECT_START);
    expect(floating_people(result).length).toEqual(start_count + 1);
    expect(floating_people(result)).toMatchSnapshot();
  });
});
