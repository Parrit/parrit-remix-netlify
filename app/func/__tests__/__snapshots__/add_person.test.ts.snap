// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add_person add to floating 1`] = `
{
  "floating": {
    "exempt": false,
    "id": "-1",
    "name": "floating_board_name_start",
    "people": [
      {
        "id": "floater_id_start",
        "name": "floater_name_start",
        "pairing_board_id": "-1",
        "type": "Person",
      },
      {
        "id": "99",
        "name": "Ms. Frizzle",
        "pairing_board_id": "brand new individual",
        "type": "Person",
      },
    ],
    "roles": [],
  },
  "id": "project_id_start",
  "name": "project_name_start",
  "pairingBoards": [
    {
      "exempt": false,
      "id": "red_id_start",
      "name": "red_start",
      "people": [
        {
          "id": "red_parrit_id",
          "name": "red_parrit",
          "pairing_board_id": "red_id_start",
          "type": "Person",
        },
      ],
      "roles": [],
    },
    {
      "exempt": false,
      "id": "blue_id_start",
      "name": "blue_start",
      "people": [],
      "roles": [
        {
          "id": "role1",
          "name": "role",
          "pairing_board_id": "blue_id_start",
          "type": "Role",
        },
      ],
    },
  ],
}
`;
